File: voice/backend/app/models/user_learning_profile.py
"""UserLearningProfile SQLAlchemy model — mirrors the Go GORM model in gather_rpg DB."""

import uuid
from datetime import datetime
from sqlalchemy import Column, Integer, String, DateTime, ForeignKey, Date
from sqlalchemy.dialects.postgresql import UUID as PGUUID, ARRAY
from sqlalchemy.orm import relationship

from app.db.database import Base


class UserLearningProfile(Base):
    __tablename__ = "user_learning_profiles"

    id = Column(PGUUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id = Column(PGUUID(as_uuid=True), ForeignKey("users.id"), nullable=False, unique=True, index=True)
    english_level = Column(String(20), nullable=False, default="beginner")
    preferred_tags = Column(ARRAY(String), nullable=False, default=[])
    weekly_score = Column(Integer, nullable=False, default=0)
    weekly_correct = Column(Integer, nullable=False, default=0)
    weekly_attempts = Column(Integer, nullable=False, default=0)
    week_start = Column(Date, nullable=True)
    current_level_xp = Column(Integer, nullable=False, default=0)
    total_xp = Column(Integer, nullable=False, default=0)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

File: gather-rpg-backend/internal/models/user_learning_profile.go
package models

import (
	"time"

	"github.com/google/uuid"
	"github.com/lib/pq"
	"gorm.io/gorm"
)

// UserLearningProfile stores the English learning profile for a user
type UserLearningProfile struct {
	ID             uuid.UUID       `gorm:"type:uuid;default:gen_random_uuid();primaryKey" json:"id"`
	UserID         uuid.UUID       `gorm:"type:uuid;not null;uniqueIndex" json:"user_id"`
	EnglishLevel   DifficultyLevel `gorm:"type:varchar(20);not null;default:'beginner'" json:"english_level"`
	PreferredTags  pq.StringArray  `gorm:"type:text[];not null;default:'{}'" json:"preferred_tags"` // e.g.: ["travel","work","communication"]
	WeeklyScore    int             `gorm:"not null;default:0" json:"weekly_score"`
	WeeklyCorrect  int             `gorm:"not null;default:0" json:"weekly_correct"`
	WeeklyAttempts int             `gorm:"not null;default:0" json:"weekly_attempts"`
	WeekStart      time.Time       `gorm:"type:date" json:"week_start"`
	CurrentLevelXP int             `gorm:"not null;default:0" json:"current_level_xp"`
	TotalXP        int             `gorm:"not null;default:0" json:"total_xp"`
	CreatedAt      time.Time       `json:"created_at"`
	UpdatedAt      time.Time       `json:"updated_at"`
	DeletedAt      gorm.DeletedAt  `gorm:"index" json:"-"`

	// Relationships
	User User `gorm:"foreignKey:UserID" json:"-"`
}
